_posts/2023-08-29-grpc-in-scala-with-fs2grpc-and-scalapb.md

```diff
@@ -0,0 +1,966 @@
+
+```proto
+...
+message OrderRequest {
+  int32 orderid = 1;
+  repeated Item items = 2;
+}
+
+message OrderReply {
+  int32 orderid = 1;
+  repeated Item items = 2;
+  double total = 3;
+}
+```
+
+Above we define two more `message` types, `OrderRequest`, and `OrderReply`. `OrderRequest` consists of `orderid` of type `int32` and `items` of type `Item`, this shows that we can nest a `message` within a `message`. The `repeated` annotation means that `items` can be repeated any number of times, in scala this becomes a `Seq` of `Item`.
+`OrderReply` also consists of an `orderid` of type `int32`, `items`, a `repeated` `Item`, and `total` of type `double`.
+
+Here's the full code:
+
+```proto
+syntax = "proto3";
+
+package com.xonal.protos;
+
+service Order {
+  rpc SendOrderStream (stream OrderRequest) returns (stream OrderReply) {}
+}
+
+message Item{
+  string name =1;
+  int32 qty = 2;
+  double amount = 3;
+}
+
+message OrderRequest {
+  int32 orderid = 1;
+  repeated Item items = 2;
+}
+
+message OrderReply {
+  int32 orderid = 1;
+  repeated Item items = 2;
+  double total = 3;
+}
+```
+
+When we compile our project and check `protobuf/target/scala-3.3.0/src_managed/main`, we'll find two sets of generated files, one under a `Fs2-grpc` folder that contains our `trait` while the rest are cases classes will be under a `Scalapb` folder. If we look at the top of each of these files, we'll find that they are all under the `com.xonal.protos.orders` package namespace that we defined in our `orders.proto` file.
+
+## 4. The gRPC server.
+Let's take a look at the `OrderFs2Grpc.scala` file generated by `Fs2-grpc`:
+
+```scala
+trait OrderFs2Grpc[F[_], A] {
+  def sendOrderStream(request: _root_.fs2.Stream[F, com.xonal.protos.orders.OrderRequest], ctx: A): _root_.fs2.Stream[F, com.xonal.protos.orders.OrderReply]
+}
+```
+We'll notice that an `OrderFs2Grpc` `trait` was created with a single function `sendOrderStream()` similar to what we defined in orders.proto, however, this function has no implementation, this is what we'll need to provide in the next.
+
+Create `OrderService.scala` in the following path `src/main/scala/com/xonal/service/OrderService.scala` and add the following code:
+
+```scala
+package com.xonal.service
+
+import com.xonal.protos.orders.*
+import cats.effect.*
+import _root_.io.grpc.*
+import fs2.Stream
+
+private class OrderService extends OrderFs2Grpc[IO, Metadata] {
+  override def sendOrderStream(
+      request: Stream[IO, OrderRequest],
+      ctx: Metadata
+  ): Stream[IO, OrderReply] = {
+    request.map { orderReq =>
+      OrderReply(
+        orderReq.orderid,
+        orderReq.items,
+        orderReq.items.map(i => i.amount).reduce(_ + _)
```

Comment:
Sum would work here, but later on when I switch to using squants Money type, it would require an implicit ordering which add more boiler plate. 
I want to have minimal changes in the last section of the article. reduce works for both.